core/lib/zksync_core/src/api_server/web3/tests/mod.rs

```diff
@@ -157,18 +210,23 @@ async fn spawn_server(
         .with_tx_sender(tx_sender, vm_barrier)
         .with_pub_sub_events(pub_sub_events_sender)
         .enable_api_namespaces(Namespace::DEFAULT.to_vec())
-        .build(stop_receiver)
+        .build(stop_receiver, api_mode)
         .await
         .expect("Failed spawning JSON-RPC server");
     (server_handles, pub_sub_events_receiver)
 }
 
 #[async_trait]
 trait HttpTest {
-    async fn test(&self, client: &HttpClient, pool: &ConnectionPool) -> anyhow::Result<()>;
+    async fn test(
+        &self,
+        client: &HttpClient,
+        pool: &ConnectionPool,
+        api_mode: APIMode,
```

Comment:
This doesn't look right. A test should know on its own which kind of the server it's run with. So, you may add the corresponding getter:

```rust
trait HttpTest {
    fn api_mode(&self) -> ApiMode;
}
```

If you want to run some test for both API modes, you can do that by having it defined like this:

```rust
struct ParametricTest(ApiMode);

#[async_trait]
impl HttpTest for ParametricTest {
    fn api_mode(&self) -> ApiMode { self.0 }

    async fn test(/* ... */) {
        // Can access `self.0` if necessary.
    } 
}

#[test_casing(2, [ApiMode::Legacy, ApiMode::Modern])
#[tokio::test]
async fn parametric_test(mode: ApiMode) {
    test_http_server(ParametricTest(mode)).await;
}
```